File: Prácticas/deepspace/src/deepspace/GameUniverse.java
package deepspace;

import java.util.ArrayList;

/**
 * Represents the game, acting as the game controller.
 * Their methods are called from Controller.java (package controller)
 * 
 * @author Miguel Ángel Fernández Gutiérrez
 * @author Sergio Quijano Rey
 */
public class GameUniverse {
    /**
     * Amount of medals necessary to win the game.
     */
    private static final int WIN = 10;
    
    /**
     * Index of the station that is currently playing.
     */
    private int currentStationIndex;
    
    /**
     * Number of turns.
     */
    private int turns;
    
    /**
     * Game dice.
     */
    private Dice dice;
    
    /**
     * Current space station.
     */
    private SpaceStation currentStation;
    
    /**
     * Set of space stations that are playing.
     */
    private ArrayList<SpaceStation> spaceStations;
    
    /**
     * Current enemy star ship.
     */
    private EnemyStarShip currentEnemy;
    
    /**
     * Game state.
     */
    private GameStateController gameState;
    
    /**
     * Whether game has a space city or not.
     */
    private boolean haveSpaceCity;
    
    // -------------------------------------------------------------------------
    // Constructors
    // -------------------------------------------------------------------------
    
    /**
     * Class initializer.
     */
    public GameUniverse() {
        currentStationIndex = -1;
        turns = 0;
        dice = new Dice();
        currentStation = null;
        spaceStations = new ArrayList<>();
        currentEnemy = null;
        gameState = new GameStateController();
        haveSpaceCity = false;
    }
    
    // -------------------------------------------------------------------------
    // Getters
    // -------------------------------------------------------------------------
    
    /**
     * Getter for gameState.
     * @return gameState
     */
    public GameState getState() {
        return gameState.getState();
    }
    
    // -------------------------------------------------------------------------
    // Setters
    // -------------------------------------------------------------------------
    
    /**
     * Discard the hangar from the current space station.
     * Discarded if gameState is INIT or AFTERCOMBAT
     */
    public void discardHangar() {
        if ( gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT )
            currentStation.discardHangar();
    }
    
    /**
     * Discard a certain shield booster from the current space station.
     * Discarded if gameState is INIT or AFTERCOMBAT
     * @param i position of the shield booster to discard
     */
    public void discardShieldBooster(int i) {
        if ( gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT )
            currentStation.discardShieldBooster(i);
    }
    
    /**
     * Discard a certain shield booster in the hangar from the current space station.
     * Discarded if gameState is INIT or AFTERCOMBAT
     * @param i position of the shield booster to discard
     */
    public void discardShieldBoosterInHangar(int i) {
        if ( gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT )
            currentStation.discardShieldBoosterInHangar(i);
    }
    
    /**
     * Discard a certain weapon from the current space station.
     * Discarded if gameState is INIT or AFTERCOMBAT
     * @param i position of the weapon to discard
     */    
    public void discardWeapon(int i) {
        if ( gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT )
            currentStation.discardWeapon(i);
    }
    
    /**
     * Discard a certain weapon in the hangar from the current space station.
     * Discarded if gameState is INIT or AFTERCOMBAT
     * @param i position of the weapon to discard
     */
    public void discardWeaponInHangar(int i) {
        if ( gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT )
            currentStation.discardWeaponInHangar(i);
    }
    
    /**
     * Mount a certain shield booster from the current space station.
     * Mounted if gameState is INIT or AFTERCOMBAT
     * @param i position of the shield booster to mount
     */
    public void mountShieldBooster(int i) {
        if ( gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT )
            currentStation.mountShieldBooster(i);
    }
    
    /**
     * Mount a certain weapon from the current space station.
     * Mounted if gameState is INIT or AFTERCOMBAT
     * @param i position of the weapon to mount
     */
    public void mountWeapon(int i) {
        if ( gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT )
            currentStation.mountWeapon(i);
    }
    
    // -------------------------------------------------------------------------
    // Game methods
    // -------------------------------------------------------------------------
    
    /**
     * Checks if the space ship that has the turn has won
     * @return true, if the space ship has won; else, otherwise
     */
    public boolean haveAWinner() {
        if ( currentStation == null )
            throw new NullPointerException("ERROR: no currentStation on GameUniverse");
        else
            if ( currentStation.getNMedals() >= WIN )
                return true;
            else
                return false;
    }
    
    /**
     * Starts a match
     * @param names collection with the names of the players
     */
    public void init(ArrayList<String> names) {
        GameState state = gameState.getState();
        
        if ( state == GameState.CANNOTPLAY ) {
            spaceStations = new ArrayList<>();
            CardDealer dealer = CardDealer.getInstance();
            
            for ( int i = 0; i < names.size(); i++ ) {
                SuppliesPackage supplies = dealer.nextSuppliesPackage();
                SpaceStation station = new SpaceStation(names.get(i), supplies);
                spaceStations.add(station);
                
                int nh = dice.initWithNHangars();
                int nw = dice.initWithNWeapons();
                int ns = dice.initWithNShields();
                
                Loot lo = new Loot(0, nw, ns, nh, 0);
                station.setLoot(lo);
            }
            
            int nPlayers = names.size();
            currentStationIndex = dice.whoStarts(nPlayers);
            currentStation = spaceStations.get(currentStationIndex);
            currentEnemy = dealer.nextEnemy();
            
            gameState.next(turns, spaceStations.size());
        }
        else{
            System.out.println("Unexpected gamestate at GameUniverse.init()");
            System.out.println("The game state is: " + gameState.getState());
            System.out.println("Nothing is done!");
        }
    }
    
    /**
     * Takes turn to the next player, if there is no pending damage
     * @return true, if the turn could be changed; else, otherwise
     */
    public boolean nextTurn() {
        GameState state = gameState.getState();
        
        if ( state == GameState.AFTERCOMBAT ) {
            boolean stationState = currentStation.validState();
            
            if ( stationState ) {
                currentStationIndex = ( currentStationIndex + 1 )
                        % spaceStations.size();
                turns++;
                
                currentStation = spaceStations.get(currentStationIndex);
                currentStation.cleanUpMountedItems();
                
                CardDealer dealer = CardDealer.getInstance();
                currentEnemy = dealer.nextEnemy();
                
                gameState.next(turns, spaceStations.size());
                
                return true;
            }
            
            return false;
        }
        
        return false;
    }
    
    /**
     * Makes combat between the space station that holds the turn and the
     * current enemy. The combat is realized if the app is on a state where
     * combat is allowed
     * @return combat result
     */
    public CombatResult combat() {
        GameState state = gameState.getState();

        if ( state == GameState.BEFORECOMBAT || state == GameState.INIT ) {
            return combat(currentStation, currentEnemy);
        } else {
            return CombatResult.NOCOMBAT;
        }
    }
    
    /**
     * Executes combat, according to game rules.
     * @param station station in combat
     * @param enemy enemy in combat
     * @return combat result
     */
    CombatResult combat(SpaceStation station, EnemyStarShip enemy) {
        // decide who starts the battle
        GameCharacter ch = dice.firstShot();
        
        boolean enemyWins;
        float fire;
        ShotResult result;
        CombatResult combatResult;
        
        if ( ch == GameCharacter.ENEMYSTARSHIP ) {  // the enemy shoots first
            // the enemy shoots to the station
            fire = enemy.fire();
            result = station.receiveShot(fire);
            
            if ( result == ShotResult.RESIST ) {
                // the station resists, therefore shoots the enemy
                fire = station.fire();
                result = enemy.receiveShot(fire);
                
                // the enemy may resist or not
                enemyWins = (result == ShotResult.RESIST);
            } else
                // the station didn't resist the shoot, therefore enemy wins
                enemyWins = true;
        } else {                                    // the station shoots first
            // we get the shot power from the station and see if the enemy resists
            fire = station.fire();
            result = enemy.receiveShot(fire);
            
            // if the station can't destroy the enemy in one shot,
            // the enemy wins directly
            enemyWins = (result == ShotResult.RESIST);
        }
        
        if ( enemyWins ) {  // enemy has won, station has a chance of escaping
            float s = station.getSpeed();
            boolean moves = dice.spaceStationMoves(s);
            
            if ( !moves ) {
                // we get the damage of the enemy and give it to the station
                Damage damage = enemy.getDamage();
                station.setPendingDamage(damage);
                
                // the station has lost
                combatResult = CombatResult.ENEMYWINS;
            } else {
                // the station manages to escape
                station.move();
                
                // the station has escaped
                combatResult = CombatResult.STATIONESCAPES;
            }
        } else {            // station has won, enemy gives it its loot
            // loot earned is given -- the transformation is catched
            Loot aLoot = enemy.getLoot();
            Transformation transform = station.setLoot(aLoot);
            
            // we check the transformation
            if ( transform == Transformation.GETEFFICIENT ) {
                makeStationEfficient();
                combatResult = CombatResult.STATIONWINSANDCONVERTS;
            } else if ( transform == Transformation.SPACECITY ) {
                createSpaceCity();
                combatResult = CombatResult.STATIONWINSANDCONVERTS;
            } else
                combatResult = CombatResult.STATIONWINS;
        }
        
        gameState.next(turns, spaceStations.size());
        
        return combatResult;
    }
    
    /**
     * Create a space city in current station.
     */
    private void createSpaceCity() {
        if ( !haveSpaceCity ) {
            ArrayList<SpaceStation> others = new ArrayList<>();
            
            for ( SpaceStation station : spaceStations )
                if ( station != currentStation )
                    others.add(station);
            
            currentStation = new SpaceCity(currentStation, others);
            spaceStations.remove(currentStationIndex);
            spaceStations.add(currentStationIndex, currentStation);
            //spaceStations.set(currentStationIndex, currentStation);
            haveSpaceCity = true;
        }
    }
    
    /**
     * Make current station efficient or beta efficient, depending on dice.
     */
    private void makeStationEfficient() {
        currentStation = new PowerEfficientSpaceStation(currentStation);
        
        if ( dice.extraEfficiency() )
            currentStation = new BetaPowerEfficientSpaceStation(currentStation);
        
        spaceStations.set(currentStationIndex, currentStation);
    }
    
    // -------------------------------------------------------------------------
    // String representation, UI version
    // -------------------------------------------------------------------------
    
    /**
     * String representation of the object.
     * @return string representation
     */
    @Override
    public String toString() {
        return  "GameUniverse(\n" +
                "\tcurrentStationIndex = " + currentStationIndex + "\n" +
                "\tcurrentStation = " + currentStation + "\n" +
                "\tcurrentEnemy = " + currentEnemy + "\n" +
                "\tturns = " + turns + "\n" +
                "\tdice = " + dice + "\n" +
                "\tgameState = " + gameState + "\n" +
                "\tspaceStations = " + spaceStations + "\n" +
                "\tWIN = " + WIN + "\n" +
                ")";
    }
    
    /**
     * To UI.
     */
    public GameUniverseToUI getUIversion() {
        return new GameUniverseToUI(currentStation, currentEnemy);
    }
    
}

File: Prácticas/Exámenes/Deepspace_Pr3_otro/lib/GameUniverse.rb
#encoding:utf-8

require_relative 'GameUniverseToUI'
require_relative 'GameStateController'
require_relative 'Dice'
require_relative 'Loot'
require_relative 'CombatResult'
require_relative 'GameCharacter'
require_relative 'ShotResult'
require_relative 'SpaceStation'
require_relative 'CardDealer'
require_relative 'EnemyStarShip'

module Deepspace

#
# @author Miguel Ángel Fernández Gutiérrez, Sergio Quijano Rey
class GameUniverse

	# Class atributes
	# ==========================================================================

	# @!attribute [Integer] amount of medals necessary to win the game
	@@WIN = 10

	# Constructors
	# ==========================================================================
	
	# Class initializer
	def initialize
		# @!attribute [GameStateController] game state
		@gameState = GameStateController.new

		# @!attribute [Integer] number of turns
		@turns = 0

		# @!attribute [Dice] game dice
		@dice = Dice.new

		# @!attribute [Integer] index of the station that is currently playing
		@currentStationIndex = -1

		# @!attribute [SpaceStation] current space station
		@currentStation = nil

		# @!attribute [Array<SpaceStation>] set of space stations that are playing
		@spaceStations = []

		# @!attribute [EnemyStarShip] current enemy star ship
		@currentEnemy = nil
	end

	# Getters
	# ==========================================================================
	
	# Getter for gameState
	# @return [GameState] gameState
	def state
		return @gameState.state
	end

	# Setters
	# ==========================================================================

	# Discard the hangar from the current space station
	# Discarded if gameState is INIT or AFTERCOMBAT
	def discardHangar
		if state == GameState::INIT or GameState::AFTERCOMBAT
			@currentStation.discardHangar
		end
	end

	# Discard a certain shield booster from the current space station
	# Discarded if gameState is INIT or AFTERCOMBAT
	# @param i [Integer] position of the shield booster to be discarded
	def discardShieldBooster(i)
		if state == GameState::INIT or GameState::AFTERCOMBAT
			@currentStation.discardShieldBooster(i)
		end
	end
	
	# Discard a certain shield booster in the hangar from the current space station.
    # Discarded if gameState is INIT or AFTERCOMBAT
    # @param i [Integer] position of the shield booster to discard
    def discardShieldBoosterInHangar(i)
		if state == GameState::INIT or GameState::AFTERCOMBAT
            @currentStation.discardShieldBoosterInHangar(i)
		end
	end
    
	# Discard a certain weapon from the current space station.
    # Discarded if gameState is INIT or AFTERCOMBAT
    # @param i [Integer] position of the weapon to discard
    def discardWeapon(i)
		if state == GameState::INIT or GameState::AFTERCOMBAT
			@currentStation.discardWeapon(i)
		end
	end
    
	# Discard a certain weapon in the hangar from the current space station.
    # Discarded if gameState is INIT or AFTERCOMBAT
    # @param i [Integer] position of the weapon to discard
    def discardWeaponInHangar(i)
		if state == GameState::INIT or GameState::AFTERCOMBAT
            @currentStation.discardWeaponInHangar(i)
		end
	end
    
    # Mount a certain shield booster from the current space station.
    # Mounted if gameState is INIT or AFTERCOMBAT
    # @param i [Integer] position of the shield booster to mount
    def mountShieldBooster(i)
		if state == GameState::INIT or GameState::AFTERCOMBAT
			@currentStation.mountShieldBooster(i)
		end
	end
	
    # Mount a certain weapon from the current space station.
    # Mounted if gameState is INIT or AFTERCOMBAT
    # @param i [Integer] position of the weapon to mount
    def mountWeapon(i)
		if state == GameState::INIT or GameState::AFTERCOMBAT
			@currentStation.mountWeapon(i)
		end
	end

	# Checks if space ship that has the turn has won
	# @return [Boolean] true, if the space ship has won; false, otherwise
	def haveAWinner
		if @currentStation.nil?
			raise "Warning! @currentStation nil on GameUniverse.haveAWinner()"
		else
			if @currentStation.nMedals >= @@WIN
				return true
			else
				return false
			end
		end
	end

	# Game methods
	# ==========================================================================

	# Starts a match
	# @param names [Array<String>] collection with the names of the players
	def init(names)
		# state = @gameState.state
		
		if state == GameState::CANNOTPLAY
			dealer = CardDealer.instance # behaviour introduced by Singleton
			names.each do |name|
				supplies = dealer.nextSuppliesPackage
				station = SpaceStation.new(name, supplies)
				@spaceStations << station

				nh = @dice.initWithNHangars
				nw = @dice.initWithNWeapons
				ns = @dice.initWithNShields

				lo = Loot.new(0, nw, ns, nh, 0)
				station.setLoot(lo)
			end

			nPlayers = names.length
			@currentStationIndex = @dice.whoStarts(nPlayers)
			@currentStation = @spaceStations[@currentStationIndex]
			@currentEnemy = dealer.nextEnemy

			@gameState.next(@turns, @spaceStations.length)
		end
	end

	# Takes turn to the next player, if there is no pending damage
	# @return [Boolean] true, if the turn could be changed; else, otherwise
	def nextTurn
		# state = @gameState.state

		if state == GameState::AFTERCOMBAT
			stationState = @currentStation.validState

			if stationState
				@currentStationIndex = ( @currentStationIndex + 1 ) \
						% @spaceStations.length
				@turns += 1

				@currentStation = @spaceStations[@currentStationIndex]
				@currentStation.cleanUpMountedItems

				dealer = CardDealer.instance # behaviour introduced by Singleton
				@currentEnemy = dealer.nextEnemy

				@gameState.next(@turns, @spaceStations.length)
				
				return true
			end

			return false
		end

		return false
	end

	# Makes combat between the space station that holds the turn and the
	# current enemy. The combat is realized if the app is on a state where
	# combat is allowed
	# @return [CombatResult] combat result
	def combat
		# state = @gameState.state

		if state == GameState::BEFORECOMBAT or GameState::INIT
			return combatGo(@currentStation, @currentEnemy)
		else
			return CombatResult::NOCOMBAT
		end
	end

	# EXAMEN
	def fireWithPenalty(station)
		if station.penalty
			if @dice.penaltySet
				fire = station.fire(@dice.generatePenalty)
				station.penalty = false
				puts "[EXAMEN] Penalización aplicada para estación"
			else
				fire = station.fire(1.0)
				puts "[EXAMEN] Penalización todavía no aplicada para estación. ¡Cuidado que viene! (je, je)"
			end
		else
			fire = station.fire(1.0)
			puts "[EXAMEN] Penalización no aplicada para estación"
		end

		return fire
	end

	# Executes combat, according to game rules
	# @param station [SpaceStation] station in combat
	# @param enemy [EnemyStarShip] enemy in combat
	# @return [CombatResult] combat result
	def combatGo(station, enemy)
		ch = @dice.firstShot

		if ch == GameCharacter::ENEMYSTARSHIP
			fire = fireWithPenalty(station)
			result = station.receiveShot(fire)

			if result == ShotResult::RESIST
				fire = fireWithPenalty(station)
				result = enemy.receiveShot(fire)

				enemyWins = (result == ShotResult::RESIST)
			else
				enemyWins = true
			end
		else
			fire = fireWithPenalty(station)
			result = enemy.receiveShot(fire)

			enemyWins = (result == ShotResult::RESIST)
		end

		if enemyWins
			s = station.speed
			moves = @dice.spaceStationMoves(s)

			if !moves
				damage = enemy.damage
				station.setPendingDamage(damage)

				combatResult = CombatResult::ENEMYWINS
			else
				station.move

				combatResult = CombatResult::STATIONESCAPES
			end
		else
			aLoot = enemy.loot
			station.setLoot(aLoot)

			combatResult = CombatResult::STATIONWINS
		end

		@gameState.next(@turns, @spaceStations.length)

		return combatResult
	end


	# String representation, UI version
	# ==========================================================================

	# String representation of the object
	# @return [String] string representation
	def to_s
		message = "[GameUniverse] -> Game state: #{@gameState.to_s}," \
				+ "Turns: #{@turns}, Dice: #{@dice.to_s}\n" \
				+ "\tCurrent station: #{@currentStation.to_s}\n" \
				+ "\tCurrent enemy: #{@currentEnemy.to_s}"
        return message
	end

	# To UI
	def getUIversion
		return EnemyToUI.new(self)
	end
	# Description:
	# 	Gets the UI representation of the object
	# Returns:
	# 	GameUniverseToUI: the UI representation
	def getUIversion
		return GameUniverseToUI.new(@currentStation, @currentEnemy)
	end
end # class GameUniverse

end # module Deepspace